File: internal/opamfile/opamfile.go
// Package opamfile parses and rewrites the downloadable sources of an opam
// package description.
//
// It deliberately implements only as much of the opam file format as a mirror
// needs: locating the top-level "url" section and every "extra-source" section,
// reading their "src" and "checksum" fields, and replacing the value of each
// "src" without disturbing a single other byte of the file.
//
// Everything else in the file is opaque, and that is on purpose. A mirror that
// re-serialises opam files it does not fully understand will eventually corrupt
// one; a mirror that splices a string literal in place cannot.
//
// Parse never fails: a file it cannot make sense of simply yields no sources,
// and a mirror leaves such a file untouched rather than crashing on it. One
// malformed package upstream must not take down a rebuild of twenty thousand.
package opamfile

import (
	"errors"
	"fmt"
	"sort"
	"strings"

	"bytes"
)

// ErrNoURL reports that a file carries no usable url.src. conf-* packages and
// virtual packages have no source archive at all, which is normal.
var ErrNoURL = errors.New("opamfile: no url.src")

// Checksum is one integrity digest declared by a package.
type Checksum struct {
	Kind string // "md5", "sha256" or "sha512"
	Hex  string
}

func (c Checksum) String() string { return c.Kind + "=" + c.Hex }

// Source is one downloadable archive a package refers to: its url section, or
// one of its extra-source sections. opam fetches extra-sources (patches and the
// like) separately at build time, so a mirror that ignores them leaves a hole
// through which builds still reach the public internet.
type Source struct {
	Name      string // "" for the url section; the label for an extra-source
	Src       string
	Checksums []Checksum

	// Byte range of the src string literal, quotes included, in the file it
	// was parsed from.
	start, end int
}

// IsExtra reports whether this is an extra-source rather than the url section.
func (s Source) IsExtra() bool { return s.Name != "" }

// File is the set of sources declared by an opam file.
type File struct {
	URL   *Source  // the package's own source archive; nil for conf/virtual packages
	Extra []Source // extra-source sections, in file order
}

// Sources returns the url section (if any) followed by every extra-source.
func (f *File) Sources() []Source {
	out := make([]Source, 0, 1+len(f.Extra))
	if f.URL != nil {
		out = append(out, *f.URL)
	}
	return append(out, f.Extra...)
}

// Parse extracts every source of an opam file. It never returns an error; a
// file with no url section yields f.URL == nil.
func Parse(data []byte) (*File, error) {
	f := &File{}
	walkSections(data, func(name, label string, bodyStart, bodyEnd int) {
		switch {
		case name == "url":
			if src, ok := parseSource(data, bodyStart, bodyEnd, ""); ok && f.URL == nil {
				f.URL = &src
			}
		case name == "extra-source" && label != "":
			if src, ok := parseSource(data, bodyStart, bodyEnd, label); ok {
				f.Extra = append(f.Extra, src)
			}
		}
	})
	return f, nil
}

// parseSource reads the src and checksum fields inside one section body.
func parseSource(data []byte, bodyStart, bodyEnd int, label string) (Source, bool) {
	src := Source{Name: label}
	s := &scanner{data: data[:bodyEnd], pos: bodyStart}
	for {
		s.skipTrivia()
		if s.eof() {
			break
		}
		name := s.readIdent()
		if name == "" {
			s.pos++ // no progress possible; do not spin
			continue
		}
		s.skipTrivia()
		if s.eof() || s.data[s.pos] != ':' {
			continue
		}
		s.pos++
		s.skipTrivia()

		switch name {
		case "src", "archive":
			if lit, start, end, ok := s.readString(); ok {
				src.Src, src.start, src.end = lit, start, end
			}
		case "checksum":
			src.Checksums = append(src.Checksums, s.readChecksums()...)
		default:
			s.skipValue()
		}
	}
	if src.Src == "" {
		return Source{}, false
	}
	return src, true
}

// RewriteSources returns a copy of data where the src of each source is replaced
// by newSrc(source). A source for which newSrc returns "" is left untouched.
// Every other byte of the file is preserved exactly.
//
// A replacement is rejected if it contains a quote, a backslash or a control
// character. Such a value is the single most likely way to silently produce an
// opam repository that parses but resolves to nothing, so it is refused here
// rather than written out and discovered by a build three weeks later.
func RewriteSources(data []byte, f *File, newSrc func(Source) string) ([]byte, error) {
	type repl struct {
		start, end int
		val        string
	}
	var repls []repl

	add := func(s Source) error {
		v := newSrc(s)
		if v == "" {
			return nil
		}
		if i := strings.IndexAny(v, "\"\\\n\r\t"); i >= 0 {
			return fmt.Errorf("opamfile: refusing to write src containing %q: %q", v[i], v)
		}
		repls = append(repls, repl{s.start, s.end, v})
		return nil
	}
	if f.URL != nil {
		if err := add(*f.URL); err != nil {
			return nil, err
		}
	}
	for _, e := range f.Extra {
		if err := add(e); err != nil {
			return nil, err
		}
	}

	if len(repls) == 0 {
		return append([]byte(nil), data...), nil
	}
	sort.Slice(repls, func(i, j int) bool { return repls[i].start < repls[j].start })

	out := make([]byte, 0, len(data)+64)
	prev := 0
	for _, r := range repls {
		out = append(out, data[prev:r.start]...)
		out = append(out, '"')
		out = append(out, r.val...)
		out = append(out, '"')
		prev = r.end
	}
	return append(out, data[prev:]...), nil
}

// RewriteSrc rewrites only the url section's src, leaving extra-sources alone.
// It is a convenience over RewriteSources for callers that touch the main
// source only.
func RewriteSrc(data []byte, newSrc string) ([]byte, error) {
	f, err := Parse(data)
	if err != nil {
		return nil, err
	}
	if f.URL == nil {
		return nil, ErrNoURL
	}
	if newSrc == "" {
		return nil, errors.New("opamfile: empty src")
	}
	return RewriteSources(data, f, func(s Source) string {
		if s.IsExtra() {
			return ""
		}
		return newSrc
	})
}

// parseChecksum reads the "kind=hex" form opam uses.
func parseChecksum(s string) (Checksum, bool) {
	kind, hex, found := strings.Cut(strings.TrimSpace(s), "=")
	if !found {
		return Checksum{}, false
	}
	switch kind {
	case "md5", "sha256", "sha512":
	default:
		return Checksum{}, false
	}
	if hex == "" {
		return Checksum{}, false
	}
	return Checksum{Kind: kind, Hex: strings.ToLower(hex)}, true
}

// --- scanner ---------------------------------------------------------------

type scanner struct {
	data []byte
	pos  int
}

func (s *scanner) eof() bool { return s.pos >= len(s.data) }

// skipTrivia consumes whitespace and comments, newlines included: opam wraps
// long values onto the following line freely.
func (s *scanner) skipTrivia() {
	for !s.eof() {
		switch c := s.data[s.pos]; {
		case c == ' ', c == '\t', c == '\n', c == '\r':
			s.pos++
		case c == '#':
			for !s.eof() && s.data[s.pos] != '\n' {
				s.pos++
			}
		default:
			return
		}
	}
}

func (s *scanner) readIdent() string {
	start := s.pos
	for !s.eof() {
		c := s.data[s.pos]
		isWord := c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' ||
			c >= '0' && c <= '9' || c == '-' || c == '_' || c == '+'
		if !isWord {
			break
		}
		s.pos++
	}
	return string(s.data[start:s.pos])
}

// readString reads a string literal and returns its unescaped value together
// with the byte range it occupies, quotes included. It handles both "..." and
// the triple-quoted """...""" form used by description fields, which is what
// keeps brace counting honest: descriptions contain braces and quotes.
func (s *scanner) readString() (val string, start, end int, ok bool) {
	if s.eof() || s.data[s.pos] != '"' {
		return "", 0, 0, false
	}
	start = s.pos

	if bytes.HasPrefix(s.data[s.pos:], []byte(`"""`)) {
		s.pos += 3
		i := bytes.Index(s.data[s.pos:], []byte(`"""`))
		if i < 0 {
			s.pos = len(s.data)
			return "", 0, 0, false
		}
		val = string(s.data[s.pos : s.pos+i])
		s.pos += i + 3
		return val, start, s.pos, true
	}

	s.pos++ // opening quote
	var b strings.Builder
	for !s.eof() {
		c := s.data[s.pos]
		switch {
		case c == '\\' && s.pos+1 < len(s.data):
			s.pos++
			switch e := s.data[s.pos]; e {
			case 'n':
				b.WriteByte('\n')
			case 't':
				b.WriteByte('\t')
			case 'r':
				b.WriteByte('\r')
			default:
				b.WriteByte(e)
			}
			s.pos++
		case c == '"':
			s.pos++
			return b.String(), start, s.pos, true
		default:
			b.WriteByte(c)
			s.pos++
		}
	}
	return "", 0, 0, false
}

// readChecksums reads either a single "kind=hex" literal or a list of them.
func (s *scanner) readChecksums() []Checksum {
	var out []Checksum
	s.skipTrivia()
	if s.eof() {
		return nil
	}

	if s.data[s.pos] != '[' {
		if lit, _, _, ok := s.readString(); ok {
			if c, ok := parseChecksum(lit); ok {
				out = append(out, c)
			}
		}
		return out
	}

	s.pos++ // '['
	for !s.eof() {
		s.skipTrivia()
		if s.eof() || s.data[s.pos] == ']' {
			s.pos++
			return out
		}
		lit, _, _, ok := s.readString()
		if !ok {
			s.pos++ // not a string; keep moving
			continue
		}
		if c, ok := parseChecksum(lit); ok {
			out = append(out, c)
		}
	}
	return out
}

// skipValue consumes one field value of any shape, plus a trailing filter.
func (s *scanner) skipValue() {
	s.skipTrivia()
	if s.eof() {
		return
	}
	switch s.data[s.pos] {
	case '"':
		s.readString()
	case '[':
		s.skipBalanced('[', ']')
	case '{':
		s.skipBalanced('{', '}')
	default:
		// A bare token: identifier, number, boolean, or a small expression
		// such as `os != "win32"`. These do not span lines.
		for !s.eof() && s.data[s.pos] != '\n' && s.data[s.pos] != '#' {
			s.pos++
		}
	}
	s.skipTrivia()
	if !s.eof() && s.data[s.pos] == '{' { // filter, e.g. {os = "linux"}
		s.skipBalanced('{', '}')
	}
}

// skipBalanced consumes a bracketed group, ignoring brackets that occur inside
// string literals or comments.
func (s *scanner) skipBalanced(open, close byte) {
	if s.eof() || s.data[s.pos] != open {
		return
	}
	depth := 0
	for !s.eof() {
		switch c := s.data[s.pos]; {
		case c == '"':
			s.readString()
			continue
		case c == '#':
			for !s.eof() && s.data[s.pos] != '\n' {
				s.pos++
			}
			continue
		case c == open:
			depth++
		case c == close:
			depth--
			if depth == 0 {
				s.pos++
				return
			}
		}
		s.pos++
	}
}

// walkSections calls fn for every top-level section, giving its name, its label
// (the quoted string for a labelled section like extra-source, "" otherwise),
// and the byte range of its body with the braces excluded.
//
// Walking the top level rather than searching for "url {" or "extra-source"
// matters: those tokens also appear inside string literals and descriptions,
// where a regex would find phantom sections.
func walkSections(data []byte, fn func(name, label string, bodyStart, bodyEnd int)) {
	s := &scanner{data: data}
	for {
		s.skipTrivia()
		if s.eof() {
			return
		}

		ident := s.readIdent()
		if ident == "" {
			s.pos++
			continue
		}
		s.skipTrivia()
		if s.eof() {
			return
		}

		switch s.data[s.pos] {
		case ':': // field
			s.pos++
			s.skipValue()
		case '{': // section
			open := s.pos
			s.skipBalanced('{', '}')
			fn(ident, "", open+1, s.pos-1)
		case '"': // labelled section: extra-source "foo" { ... }
			label, _, _, _ := s.readString()
			s.skipTrivia()
			if !s.eof() && s.data[s.pos] == '{' {
				open := s.pos
				s.skipBalanced('{', '}')
				fn(ident, label, open+1, s.pos-1)
			}
		default:
			s.pos++
		}
	}
}

File: ocaml/lib/opamfile.ml
(* Parse and rewrite the downloadable sources of an opam package file.

   This is the OCaml port of the Go package of the same name. It implements
   only as much of the opam file format as a mirror needs: the top-level [url]
   section and every [extra-source] section, their [src] and [checksum], and
   replacing each [src] in place without disturbing another byte.

   Parse never fails. A file it cannot make sense of yields no sources, and a
   mirror leaves such a file untouched rather than choking on it: one malformed
   package upstream must not take down a rebuild of twenty thousand. *)

type checksum = { kind : string; hex : string }

let checksum_to_string c = c.kind ^ "=" ^ c.hex

(* One downloadable archive: the url section, or one extra-source. opam fetches
   extra-sources (patches and the like) separately at build time, so a mirror
   that ignores them leaves a hole through which builds still reach the net. *)
type source = {
  name : string;             (* "" for the url section; the label of an extra-source *)
  src : string;
  checksums : checksum list;
  start_ : int;              (* byte range of the src literal, quotes included *)
  end_ : int;
}

let is_extra (s : source) = s.name <> ""

type t = { url : source option; extra : source list }

let sources (t : t) : source list =
  (match t.url with Some u -> [ u ] | None -> []) @ t.extra

(* --- scanner ------------------------------------------------------------- *)

type scanner = { data : string; hi : int; mutable pos : int }

let eof s = s.pos >= s.hi
let peek s = s.data.[s.pos]

let is_word = function
  | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9' | '-' | '_' | '+' -> true
  | _ -> false

(* consume whitespace and comments, newlines included: opam wraps long values
   onto the following line freely. *)
let skip_trivia s =
  let go = ref true in
  while !go && not (eof s) do
    match peek s with
    | ' ' | '\t' | '\n' | '\r' -> s.pos <- s.pos + 1
    | '#' ->
        while (not (eof s)) && peek s <> '\n' do
          s.pos <- s.pos + 1
        done
    | _ -> go := false
  done

let read_ident s =
  let start = s.pos in
  while (not (eof s)) && is_word (peek s) do
    s.pos <- s.pos + 1
  done;
  String.sub s.data start (s.pos - start)

let has_prefix_at s i lit =
  i + String.length lit <= s.hi && String.sub s.data i (String.length lit) = lit

(* read a string literal, returning its unescaped value and the byte range it
   occupies, quotes included. Handles both "..." and the triple-quoted form
   used by description fields, which is what keeps brace counting honest. *)
let read_string s : (string * int * int) option =
  if eof s || peek s <> '"' then None
  else begin
    let start = s.pos in
    if has_prefix_at s s.pos "\"\"\"" then begin
      s.pos <- s.pos + 3;
      let rec find i =
        if i + 3 > s.hi then None
        else if String.sub s.data i 3 = "\"\"\"" then Some i
        else find (i + 1)
      in
      match find s.pos with
      | None ->
          s.pos <- s.hi;
          None
      | Some i ->
          let v = String.sub s.data s.pos (i - s.pos) in
          s.pos <- i + 3;
          Some (v, start, s.pos)
    end
    else begin
      s.pos <- s.pos + 1;
      (* opening quote *)
      let b = Buffer.create 32 in
      let result = ref None in
      let go = ref true in
      while !go && not (eof s) do
        let c = peek s in
        if c = '\\' && s.pos + 1 < s.hi then begin
          s.pos <- s.pos + 1;
          (match s.data.[s.pos] with
          | 'n' -> Buffer.add_char b '\n'
          | 't' -> Buffer.add_char b '\t'
          | 'r' -> Buffer.add_char b '\r'
          | e -> Buffer.add_char b e);
          s.pos <- s.pos + 1
        end
        else if c = '"' then begin
          s.pos <- s.pos + 1;
          result := Some (Buffer.contents b, start, s.pos);
          go := false
        end
        else begin
          Buffer.add_char b c;
          s.pos <- s.pos + 1
        end
      done;
      !result
    end
  end

let parse_checksum str : checksum option =
  let str = String.trim str in
  match String.index_opt str '=' with
  | None -> None
  | Some i ->
      let kind = String.sub str 0 i in
      let hex = String.sub str (i + 1) (String.length str - i - 1) in
      (match kind with
      | "md5" | "sha256" | "sha512" ->
          if hex = "" then None
          else Some { kind; hex = String.lowercase_ascii hex }
      | _ -> None)

(* read either a single "kind=hex" literal or a bracketed list of them. *)
let read_checksums s : checksum list =
  skip_trivia s;
  if eof s then []
  else if peek s <> '[' then
    match read_string s with
    | Some (lit, _, _) -> ( match parse_checksum lit with Some c -> [ c ] | None -> [] )
    | None -> []
  else begin
    s.pos <- s.pos + 1;
    (* '[' *)
    let acc = ref [] in
    let go = ref true in
    while !go && not (eof s) do
      skip_trivia s;
      if eof s || peek s = ']' then begin
        if not (eof s) then s.pos <- s.pos + 1;
        go := false
      end
      else
        match read_string s with
        | None -> s.pos <- s.pos + 1
        | Some (lit, _, _) -> (
            match parse_checksum lit with Some c -> acc := c :: !acc | None -> ())
    done;
    List.rev !acc
  end

let rec skip_balanced s op cl =
  if (not (eof s)) && peek s = op then begin
    let depth = ref 0 in
    let go = ref true in
    while !go && not (eof s) do
      let c = peek s in
      if c = '"' then ignore (read_string s)
      else begin
        if c = '#' then
          while (not (eof s)) && peek s <> '\n' do
            s.pos <- s.pos + 1
          done
        else begin
          if c = op then incr depth
          else if c = cl then begin
            decr depth;
            if !depth = 0 then begin
              s.pos <- s.pos + 1;
              go := false
            end
          end;
          if !go then s.pos <- s.pos + 1
        end
      end
    done
  end

and skip_value s =
  skip_trivia s;
  if not (eof s) then begin
    (match peek s with
    | '"' -> ignore (read_string s)
    | '[' -> skip_balanced s '[' ']'
    | '{' -> skip_balanced s '{' '}'
    | _ ->
        (* a bare token: identifier, number, boolean or a small expression such
           as [os != "win32"]; these do not span lines. *)
        while (not (eof s)) && peek s <> '\n' && peek s <> '#' do
          s.pos <- s.pos + 1
        done);
    skip_trivia s;
    if (not (eof s)) && peek s = '{' then skip_balanced s '{' '}'
  end

(* call [fn name label body_start body_end] for every top-level section. Walking
   the top level rather than searching for "url {" matters: those tokens also
   appear inside strings and descriptions, where a regex finds phantoms. *)
let walk_sections data fn =
  let s = { data; hi = String.length data; pos = 0 } in
  let go = ref true in
  while !go do
    skip_trivia s;
    if eof s then go := false
    else begin
      let ident = read_ident s in
      if ident = "" then s.pos <- s.pos + 1
      else begin
        skip_trivia s;
        if eof s then go := false
        else
          match peek s with
          | ':' ->
              s.pos <- s.pos + 1;
              skip_value s
          | '{' ->
              let op = s.pos in
              skip_balanced s '{' '}';
              fn ident "" (op + 1) (s.pos - 1)
          | '"' -> (
              match read_string s with
              | Some (label, _, _) ->
                  skip_trivia s;
                  if (not (eof s)) && peek s = '{' then begin
                    let op = s.pos in
                    skip_balanced s '{' '}';
                    fn ident label (op + 1) (s.pos - 1)
                  end
              | None -> s.pos <- s.pos + 1)
          | _ -> s.pos <- s.pos + 1
      end
    end
  done

let parse_source data body_start body_end label : source option =
  let s = { data; hi = body_end; pos = body_start } in
  let src = ref "" and rng = ref (0, 0) and sums = ref [] in
  let go = ref true in
  while !go do
    skip_trivia s;
    if eof s then go := false
    else begin
      let name = read_ident s in
      if name = "" then s.pos <- s.pos + 1
      else begin
        skip_trivia s;
        if eof s || peek s <> ':' then ()
        else begin
          s.pos <- s.pos + 1;
          skip_trivia s;
          match name with
          | "src" | "archive" -> (
              match read_string s with
              | Some (lit, a, b) ->
                  src := lit;
                  rng := (a, b)
              | None -> ())
          | "checksum" -> sums := !sums @ read_checksums s
          | _ -> skip_value s
        end
      end
    end
  done;
  if !src = "" then None
  else
    let a, b = !rng in
    Some { name = label; src = !src; checksums = !sums; start_ = a; end_ = b }

let parse (data : string) : t =
  let url = ref None and extra = ref [] in
  walk_sections data (fun name label body_start body_end ->
      match name with
      | "url" -> (
          if !url = None then
            match parse_source data body_start body_end "" with
            | Some s -> url := Some s
            | None -> ())
      | "extra-source" when label <> "" -> (
          match parse_source data body_start body_end label with
          | Some s -> extra := s :: !extra
          | None -> ())
      | _ -> ());
  { url = !url; extra = List.rev !extra }

exception Rewrite_error of string

let has_forbidden v =
  let bad = ref false in
  String.iter
    (fun c -> match c with '"' | '\\' | '\n' | '\r' | '\t' -> bad := true | _ -> ())
    v;
  !bad

(* Return a copy of [data] where the src of each source is replaced by the value
   [replace] returns (Some v), leaving alone the ones it maps to None. Every
   other byte is preserved. A replacement carrying a quote, backslash or control
   character is refused: it is the surest way to silently produce a repository
   that parses but resolves to nothing. *)
let rewrite_sources (data : string) (t : t) (replace : source -> string option) :
    string =
  let repls = ref [] in
  let add (s : source) =
    match replace s with
    | None -> ()
    | Some v ->
        if v = "" then ()
        else if has_forbidden v then
          raise (Rewrite_error (Printf.sprintf "refusing to write src %S" v))
        else repls := (s.start_, s.end_, v) :: !repls
  in
  (match t.url with Some u -> add u | None -> ());
  List.iter add t.extra;
  match !repls with
  | [] -> data
  | repls ->
      let repls = List.sort (fun (a, _, _) (b, _, _) -> compare a b) repls in
      let b = Buffer.create (String.length data + 64) in
      let prev = ref 0 in
      List.iter
        (fun (start_, end_, v) ->
          Buffer.add_substring b data !prev (start_ - !prev);
          Buffer.add_char b '"';
          Buffer.add_string b v;
          Buffer.add_char b '"';
          prev := end_)
        repls;
      Buffer.add_substring b data !prev (String.length data - !prev);
      Buffer.contents b

(* Rewrite only the url section's src, leaving extra-sources alone. *)
let rewrite_src (data : string) (new_src : string) : string =
  let t = parse data in
  match t.url with
  | None -> raise (Rewrite_error "no url.src")
  | Some _ ->
      if new_src = "" then raise (Rewrite_error "empty src");
      rewrite_sources data t (fun s -> if is_extra s then None else Some new_src)
